apps/server/lib/lexical/server/code_intelligence/completion/translations/module_or_behaviour.ex

```diff
@@ -20,42 +20,54 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.ModuleOrBehavi
   end
 
   defp do_translate(%_{} = module, builder, %Env{} = env) do
-    struct_reference? = Env.in_context?(env, :struct_reference)
-
-    defines_struct? = Intelligence.defines_struct?(env.project, module.full_name)
-
-    immediate_descendent_structs =
-      immediate_descendent_struct_modules(env.project, module.full_name)
-
-    defines_struct_in_descendents? =
-      immediate_descendent_defines_struct?(env.project, module.full_name) and
-        length(immediate_descendent_structs) > 1
+    if Env.in_context?(env, :struct_reference) do
+      complete_in_struct_reference(env, builder, module)
+    else
+      detail = builder.fallback(module.summary, module.name)
+      completion(env, builder, module.name, detail)
+    end
+  end
 
-    cond do
-      struct_reference? and defines_struct_in_descendents? and defines_struct? ->
-        more = length(immediate_descendent_structs) - 1
+  defp complete_in_struct_reference(%Env{} = env, builder, module) do
+    ancestors_and_more =
+      env.project
+      |> immediate_descendent_struct_modules(module.full_name)
+      |> ancestors_and_more()
 
-        [
-          Translations.Struct.completion(env, builder, module.name, module.full_name, more),
-          Translations.Struct.completion(env, builder, module.name, module.full_name)
-        ]
+    Enum.flat_map(ancestors_and_more, fn {child_module_name, more} ->
+      local_name = local_module_name(module.full_name, child_module_name, module.name)
 
-      struct_reference? and defines_struct? ->
-        Translations.Struct.completion(env, builder, module.name, module.full_name)
+      env
+      |> Translations.Struct.completion(builder, local_name, child_module_name, more)
+      |> List.wrap()
+    end)
+  end
 
-      struct_reference? and
-          immediate_descendent_defines_struct?(env.project, module.full_name) ->
-        Enum.map(immediate_descendent_structs, fn child_module_name ->
-          local_name = local_module_name(module.full_name, child_module_name)
-          Translations.Struct.completion(env, builder, local_name, child_module_name)
+  defp ancestors_and_more(modules) do
+    ancestors_and_descentents =
+      for ancestor <- modules do
+        Enum.reduce(modules, {ancestor, 0, []}, fn maybe_descendant,
+                                                   {ancestor, count, descendents} ->
+          if descendent?(maybe_descendant, ancestor) do
+            {ancestor, count + 1, [maybe_descendant | descendents]}
+          else
+            {ancestor, count, descendents}
+          end
         end)
+      end
+
+    descendents =
+      Enum.flat_map(ancestors_and_descentents, fn {_, _, descendents} -> descendents end)
 
-      true ->
-        detail = builder.fallback(module.summary, module.name)
-        completion(env, builder, module.name, detail)
+    for {ancestor, count, _} <- ancestors_and_descentents, ancestor not in descendents do
```

Comment:
I'm totally lost here; why did you do everything in "one" loop (it's really not one loop, but a reduce in a comprehension, which is m*n) ?